crypto/bp/bp_g2.c

```diff
@@ -0,0 +1,1212 @@
+
+    if (!b->Z_is_one) {
+        if (!FP2_sqr(group, Zb23, b->Z, ctx))
+            goto end;
+        if (!FP2_mul(group, tmp1, a->X, Zb23, ctx))
+            goto end;
+        tmp1_ = tmp1;
+    } else
+        tmp1_ = a->X;
+    if (!a->Z_is_one) {
+        if (!FP2_sqr(group, Za23, a->Z, ctx))
+            goto end;
+        if (!FP2_mul(group, tmp2, b->X, Za23, ctx))
+            goto end;
+        tmp2_ = tmp2;
+    } else
+        tmp2_ = b->X;
+
+    /*
+     * compare X_a*Z_b^2 with X_b*Z_a^2
+     */
+    if (FP2_cmp(tmp1_, tmp2_) != 0) {
+        ret = 1;                /* points differ */
+        goto end;
+    }
+
+    if (!b->Z_is_one) {
+        if (!FP2_mul(group, Zb23, Zb23, b->Z, ctx))
+            goto end;
+        if (!FP2_mul(group, tmp1, a->Y, Zb23, ctx))
+            goto end;
+        /*
+         * tmp1_ = tmp1
+         */
+    } else {
+        tmp1_ = a->Y;
+    }
+    if (!a->Z_is_one) {
+        if (!FP2_mul(group, Za23, Za23, a->Z, ctx))
+            goto end;
+        if (!FP2_mul(group, tmp2, b->Y, Za23, ctx))
+            goto end;
+        /*
+         * tmp2_ = tmp2
+         */
+    } else {
+        tmp2_ = b->Y;
+    }
+
+    /*
+     * compare Y_a*Z_b^3 with Y_b*Z_a^3
+     */
+    if (FP2_cmp(tmp1_, tmp2_) != 0) {
+        ret = 1;                /* points differ */
+        goto end;
+    }
+
+    /*
+     * points are equal
+     */
+    ret = 0;
+
+ end:
+    FP2_free(tmp1);
+    FP2_free(tmp2);
+    FP2_free(Za23);
+    FP2_free(Zb23);
+    BN_CTX_free(new_ctx);
+    return ret;
+}
+
+int G2_ELEM_make_affine(const BP_GROUP *group, G2_ELEM *point, BN_CTX *ctx)
+{
+    BN_CTX *new_ctx = NULL;
+    BIGNUM *x[2], *y[2];
+    int ret = 0;
+
+    if (point->Z_is_one || G2_ELEM_is_at_infinity(group, point))
+        return 1;
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_new()) == NULL)
+        return 0;
+
+    BN_CTX_start(ctx);
+    if ((x[0] = BN_CTX_get(ctx)) == NULL || (x[1] = BN_CTX_get(ctx)) == NULL
+        || (y[0] = BN_CTX_get(ctx)) == NULL
+        || (y[1] = BN_CTX_get(ctx)) == NULL)
+        goto err;
+
+    if (!G2_ELEM_get_affine_coordinates(group, point, x, y, ctx))
+        goto err;
+    if (!G2_ELEM_set_affine_coordinates
+        (group, point, (const BIGNUM **)x, (const BIGNUM **)y, ctx))
+        goto err;
+    if (!point->Z_is_one)
+        goto err;
+    ret = 1;
+
+ err:
+    BN_CTX_end(ctx);
+    BN_CTX_free(new_ctx);
+    return ret;
+}
+
+int G2_ELEMs_make_affine(const BP_GROUP *group, size_t num,
+                         G2_ELEM *points[], BN_CTX *ctx)
+{
+    BN_CTX *new_ctx = NULL;
+    FP2 *zs[num], *z2 = NULL, *z3 = NULL;
+    size_t i, m = 0;
+    int ret = 0;
+
+    if (num == 0)
+        return 1;
+
+    if (num == 1)
+        return G2_ELEM_make_affine(group, points[0], ctx);
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_new()) == NULL)
+        return 0;
+
+    if ((z2 = FP2_new()) == NULL || (z3 = FP2_new()) == NULL)
+        goto err;
+
+    m = 0;
```

Comment:
`m` counts the number of non-trivial field elements so simultaneously invert, but I removed the double initialization.